### PR TITLE
sql: reduce unnecessary latency in `SplitTable` for test

### PR DESCRIPTION
**Pull request by hueypark**
Before this patch, test waits for the replication to complete on all destination nodes when each `SplitTable` called. The idea let the all `SplitTable` call once and wait only once after that. This makes the tests faster and detail result(in my local machine) is below.

| func | before | after |
| ---- | ------ | ----- |
| TestDistBackfill | 5 s 230 ms | ~~910 ms~~ 1 s 980 ms |
| TestRaceWithBackfill | 4 s 770 ms | ~~1 s 620 ms~~ 2 s 540 ms |
| TestDropWhileBackfill | 3 s 960 ms | ~~790 ms~~ 1 s 800 ms |
| TestBackfillErrors | 4 s 940 ms | ~~750 ms~~ 1 s 690 m |
| TestBackfillCompletesOnChunkBoundary | 5 s 780 ms | ~~1 s 590 ms~~ 2 s 640 ms |
| TestSchemaChangeEvalContext | 2 s 870 ms | ~~740 ms~~ 1 s 970 ms |
| TestCancelSchemaChange | 5 s 650 ms | ~~1 s 500 ms~~ 2 s 420 ms |
| TestIntentRaceWithIndexBackfill | 330 ms | 310 ms |

Related to #37028.

Release note: None

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
This is a very nice find, thanks!

@jordanlewis I'm not super familiar with the details of the test cluster but I'm not sure who should review (my guess would be @andreimatei but he's out for the week).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-:-LkuU2wL5w1obR7uuKiV:b-j29ar7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/distsql_physical_planner_test.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSXee26ilmR75PDwO:-LkuSXee26ilmR75PDwP:b-tz5uhj) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L52)):*
> ```Go
> // side of the split on targetNodeIdx, and moves the lease for the right
> // side of the split to targetNodeIdx. This forces the querying against
> // the table to be distributed. vals is a list of values forming a primary
> ```

This comment needs updating (the sentence about `vals`)
___
*[pkg/sql/distsql_physical_planner_test.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSS0R6baFTrNcVx7f:-LkuSS0R6baFTrNcVx7g:bohcwbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L63)):*
> ```Go
> 	tc serverutils.TestClusterInterface,
> 	desc *sqlbase.TableDescriptor,
> 	argss []SplitTableArgs,
> ```

[nit] `argss` needs a better name, perhaps `splitPoints` (same in other places)
___
*[pkg/sql/distsql_physical_planner_test.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuRw2c4J0VwuwIY-Fv:-LkuRw2c4J0VwuwIY-Fw:bdbrsea) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L106)):*
> ```Go
> 
> // SplitTableArgs contains the parameters one can SplitTable.
> type SplitTableArgs struct {
> ```

[nit] a better name would be `SplitPoint`. The comment can be better, e.g. "SplitPoint describes a split point that is passed to SplitTable."
___
*[pkg/sql/distsql_physical_planner_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuRyCJ9amM2zc354HF:-LkuRyCJ9amM2zc354HG:bf9sch6) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L107)):*
> ```Go
> // SplitTableArgs contains the parameters one can SplitTable.
> type SplitTableArgs struct {
> 	TargetNodeIdx int
> ```

The fields should have comments. "TargetNodeIdx is the node that will have the lease for the new range" 
___
*[pkg/sql/distsql_physical_planner_test.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuS74_0wPEPY92-M1M:-LkuS74_0wPEPY92-M1N:b-4lv9pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L108)):*
> ```Go
> type SplitTableArgs struct {
> 	TargetNodeIdx int
> 	Val           int
> ```

This should be called something more informative like `PK` and it should be a `[]interface{}` for generality.
___
*[pkg/sql/distsql_plan_backfill_test.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSH5s6zkw7Am7j406:-LkuSH5s6zkw7Am7j407:b-oktmpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_plan_backfill_test.go#L78)):*
> ```Go
> 	// Split the table into multiple ranges.
> 	descNumToStr := sqlbase.GetTableDescriptor(cdb, "test", "numtostr")
> 	// SplitTable moves the right range, so we split things back to front
> ```

Does this back-to-front thing still make sense now if we're doing them in parallel?
___
*[pkg/testutils/serverutils/test_cluster_shim.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuTA-E7ck0gASGaUm-:-LkuTA-E7ck0gASGaUm0:b-yxwwk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/serverutils/test_cluster_shim.go#L49)):*
> ```Go
> 	// It's illegal to have multiple replicas of the same range on stores of a single
> 	// node.
> 	// The method blocks until a snapshot of the range has been copied to all the
> ```

This description is not true anymore, is it? We no longer wait.. This could break existing uses of this method.
Maybe we want to have two versions, or pass a flag.
___
*[pkg/testutils/serverutils/test_cluster_shim.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuShSx-rAiIZzZ1bie:-LkuShSx-rAiIZzZ1bif:bhfx82y) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/serverutils/test_cluster_shim.go#L61)):*
> ```Go
> 	) roachpb.RangeDescriptor
> 
> 	// WaitForReplicationComplete wait for replication complete.
> ```

This needs a better description. What is the start key for, and what are the targets for?
___
*[pkg/testutils/testcluster/testcluster.go, line 404 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSqVq0CyuTpq6uSEI:-LkuSqVq0CyuTpq6uSEJ:b-40olcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/testcluster/testcluster.go#L404)):*
> ```Go
> }
> 
> // WaitForReplicationComplete wait for replication complete.
> ```

[nit] This should just be `WaitForReplicationComplete is part of TestClusterInterface.` (so we don't duplicate the description).

**Review by andreimatei — CHANGES_REQUESTED** [review]
Thanks for the patch!

**Review by hueypark — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/distsql_physical_planner_test.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSXee26ilmR75PDwO:-LlHY3nz7gUlDOeK1Biq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L52)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This comment needs updating (the sentence about `vals`)
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner_test.go, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSS0R6baFTrNcVx7f:-LlHYCaMAjKtTKaJNJp-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L63)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] `argss` needs a better name, perhaps `splitPoints` (same in other places)
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner_test.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuRw2c4J0VwuwIY-Fv:-LlHYKWF-q3-w7jDswXG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L106)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] a better name would be `SplitPoint`. The comment can be better, e.g. "SplitPoint describes a split point that is passed to SplitTable."
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner_test.go, line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuRyCJ9amM2zc354HF:-LlHYM803xmOgc4zlsUk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L107)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The fields should have comments. "TargetNodeIdx is the node that will have the lease for the new range" 
</blockquote></details>

Done.
___
*[pkg/sql/distsql_physical_planner_test.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuS74_0wPEPY92-M1M:-LlHYONr5lsJCVOL3n_0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_physical_planner_test.go#L108)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should be called something more informative like `PK` and it should be a `[]interface{}` for generality.
</blockquote></details>

Done.
___
*[pkg/sql/distsql_plan_backfill_test.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuSH5s6zkw7Am7j406:-LlHYSzGB84VNT_BpRJp:b9bgep2) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/sql/distsql_plan_backfill_test.go#L78)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Does this back-to-front thing still make sense now if we're doing them in parallel?
</blockquote></details>

Done. This back-to-front thing does not make sense now.
___
*[pkg/testutils/serverutils/test_cluster_shim.go, line 49 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuTA-E7ck0gASGaUm-:-LlHYf7-5i0DsGUrZy_t:b3bnbuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/serverutils/test_cluster_shim.go#L49)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This description is not true anymore, is it? We no longer wait.. This could break existing uses of this method.
Maybe we want to have two versions, or pass a flag.
</blockquote></details>

I create two versions.
___
*[pkg/testutils/serverutils/test_cluster_shim.go, line 61 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkuShSx-rAiIZzZ1bie:-LlHYlEP1grKJKhW1bln:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/serverutils/test_cluster_shim.go#L61)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This needs a better description. What is the start key for, and what are the targets for?
</blockquote></details>

Done.

File: pkg/testutils/testcluster/testcluster.go
Comment (hueypark) [reply pkg/testutils/testcluster/testcluster.go]: I create two versions for that. Now all AddReplicas related functions are waiting again.

File: pkg/testutils/testcluster/testcluster.go
Comment (hueypark) [reply pkg/testutils/testcluster/testcluster.go]: I remove unnecessary for loop and erase "likely downreplicated again". @tbg can you review this?

**Comment by hueypark** [comment]
@RaduBerinde @andreimatei Thank you for the review!
The first attempt was a problem (transfer range lease before replication completed). Please note that the problem has now been fixed.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @hueypark, @jordanlewis, @RaduBerinde, and @tbg)

___
*[pkg/sql/distsql_physical_planner_test.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LlcXK4yAK7dGuMojwWJ:-LlcXK4z516fYyN6RbvY:b-z1d2wx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e13ab4190063dd7797b0787b93407a1f485e02/pkg/sql/distsql_physical_planner_test.go#L50)):*
> ```Go
> 
> // SplitTable splits a range in the table, creates a replica for the right
> // side of the split on targetNodeIdx, and moves the lease for the right
> ```

comment is now stale
___
*[pkg/testutils/testcluster/testcluster.go, line 411 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkrViBUx01JWlMohUqW-r1-411:-LlcYa6T6mCCfTgP9zZC:bn3nm24) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/testcluster/testcluster.go#L411)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

I remove unnecessary for loop and erase "likely downreplicated again". @tbg can you review this?
</blockquote></details>

Tobi was on vacation (as was I), but we're back now. Friendly ping.

**Review by hueypark — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, @RaduBerinde, and @tbg)

___
*[pkg/sql/distsql_physical_planner_test.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LlcXK4yAK7dGuMojwWJ:-Lld6vvnAes3B4sgw8aP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e13ab4190063dd7797b0787b93407a1f485e02/pkg/sql/distsql_physical_planner_test.go#L50)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

comment is now stale
</blockquote></details>

Done.

**Comment by hueypark** [comment]
I found `TestStoreMetrics` is failed(https://teamcity.cockroachdb.com/viewLog.html?buildId=1426234) in CI. But I can't reproduce this fail. Could anyone explain this fail?

**Comment by andreimatei** [comment]
There's talk of that test here:
https://github.com/cockroachdb/cockroach/issues/39390

**Comment by hueypark** [comment]
@andreimatei Thanks!

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, and @RaduBerinde)

___
*[pkg/testutils/testcluster/testcluster.go, line 411 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkrViBUx01JWlMohUqW-r1-411:-LllgVSp9Bz5h4GP3Yci:b-7ve3vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/testcluster/testcluster.go#L411)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Tobi was on vacation (as was I), but we're back now. Friendly ping.
</blockquote></details>

Here's what past Tobias had to say about this:

https://github.com/cockroachdb/cockroach/commit/4b20a079c78cbc2ed13f1caa38e5cc6938ff08bc

Present Tobias doesn't know if that is still a problem, but if the replicate queue were active, it would be reasonable to assume that it could undo some work we've done here. I don't know that there's a compelling reason for wanting to move replicas around when the queue *is* active, so maybe that should just be a hard error right there? And then this code can certainly go.

**Review by andreimatei — APPROVED** [review]
LGTM

Just some nits about a method name remaining. Once you address them, I'll merge this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-:-Llll2JtE1Sj1Y4EIKxd:blphsuv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @hueypark, @jordanlewis, and @RaduBerinde)

___
*[pkg/testutils/testcluster/testcluster.go, line 411 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LkrViBUx01JWlMohUqW-r1-411:-Lllkuze67tKSCKRuP-2:b-kcjjv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/de8043e0ffe995f9f9a933e426195085cee2f6fb/pkg/testutils/testcluster/testcluster.go#L411)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Here's what past Tobias had to say about this:

https://github.com/cockroachdb/cockroach/commit/4b20a079c78cbc2ed13f1caa38e5cc6938ff08bc

Present Tobias doesn't know if that is still a problem, but if the replicate queue were active, it would be reasonable to assume that it could undo some work we've done here. I don't know that there's a compelling reason for wanting to move replicas around when the queue *is* active, so maybe that should just be a hard error right there? And then this code can certainly go.
</blockquote></details>

OK, I say screw the loop then. If tests start flaky, we'll consider again how to deal with them.
___
*[pkg/testutils/testcluster/testcluster.go, line 450 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LllhPmJ9LZKo9dTMrKj:-LllhPmJ9LZKo9dTMrKk:bbepie5) ([raw file](https://github.com/cockroachdb/cockroach/blob/573043fc5004aa383d607ee3beb0fe72d8719777/pkg/testutils/testcluster/testcluster.go#L450)):*
> ```Go
> }
> 
> // waitForReplicationComplete waits replication complete.
> ```

nit: waits *for* replication *to* complete.

And maybe a better name would be `waitForNewReplicas()` and the comment can say that the method waits for each of the targets to have a fully initialized replica of the range indicated by startKey.

**Review by hueypark — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, and @RaduBerinde)

___
*[pkg/testutils/testcluster/testcluster.go, line 450 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-LllhPmJ9LZKo9dTMrKj:-Llm_OO94chZ4SY3BVN9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/573043fc5004aa383d607ee3beb0fe72d8719777/pkg/testutils/testcluster/testcluster.go#L450)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: waits *for* replication *to* complete.

And maybe a better name would be `waitForNewReplicas()` and the comment can say that the method waits for each of the targets to have a fully initialized replica of the range indicated by startKey.
</blockquote></details>

Done.

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39142#-:-LlnIG_b1tbm1rhCUQY9:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, and @RaduBerinde)

**Pull request merged**
